File: heroic-dist/src/test/java/com/spotify/heroic/LoggingMetricModule.java
package com.spotify.heroic;

import com.spotify.heroic.async.AsyncObservable;
import com.spotify.heroic.common.Groups;
import com.spotify.heroic.common.Statistics;
import com.spotify.heroic.dagger.PrimaryComponent;
import com.spotify.heroic.instrumentation.OperationsLog;
import com.spotify.heroic.lifecycle.LifeCycle;
import com.spotify.heroic.metric.BackendEntry;
import com.spotify.heroic.metric.BackendKey;
import com.spotify.heroic.metric.BackendKeyFilter;
import com.spotify.heroic.metric.BackendKeySet;
import com.spotify.heroic.metric.FetchData;
import com.spotify.heroic.metric.FetchQuotaWatcher;
import com.spotify.heroic.metric.MetricBackend;
import com.spotify.heroic.metric.MetricCollection;
import com.spotify.heroic.metric.MetricModule;
import com.spotify.heroic.metric.WriteMetric;
import eu.toolchain.async.AsyncFuture;
import java.util.List;
import java.util.function.Consumer;
import lombok.Data;

@Data
class LoggingMetricModule implements MetricModule {
    private final MetricModule backingStore;
    private final OperationsLog opLog;

    @Override
    public Exposed module(
        final PrimaryComponent primary, final Depends backend, final String id
    ) {
        final Exposed backingStoreModule = backingStore.module(primary, backend, id);

        return new Exposed() {

            @Override
            public LifeCycle life() {
                return backingStoreModule.life();
            }

            @Override
            public MetricBackend backend() {
                final MetricBackend delegate = backingStoreModule.backend();

                return new LoggingMetricBackend(delegate);
            }
        };
    }

    @Data
    private class LoggingMetricBackend implements MetricBackend {
        private final MetricBackend delegate;

        @Override
        public Statistics getStatistics() {
            return delegate.getStatistics();
        }

        @Override
        public AsyncFuture<Void> configure() {
            return delegate.configure();
        }

        @Override
        public AsyncFuture<WriteMetric> write(
            final WriteMetric.Request write
        ) {
            opLog.registerWriteRequest();
            return delegate.write(write).directTransform(result -> {
                opLog.registerWriteComplete();
                return result;
            });
        }

        @Override
        public AsyncFuture<FetchData> fetch(
            final FetchData.Request request, final FetchQuotaWatcher watcher
        ) {
            return delegate.fetch(request, watcher);
        }

        @Override
        public AsyncFuture<FetchData.Result> fetch(
            final FetchData.Request request, final FetchQuotaWatcher watcher,
            final Consumer<MetricCollection> metricsConsumer
        ) {
            return delegate.fetch(request, watcher, metricsConsumer);
        }

        @Override
        public Iterable<BackendEntry> listEntries() {
            return delegate.listEntries();
        }

        @Override
        public AsyncObservable<BackendKeySet> streamKeys(
            final BackendKeyFilter filter, final QueryOptions options
        ) {
            return delegate.streamKeys(filter, options);
        }

        @Override
        public AsyncObservable<BackendKeySet> streamKeysPaged(
            final BackendKeyFilter filter, final QueryOptions options, final long pageSize
        ) {
            return delegate.streamKeysPaged(filter, options, pageSize);
        }

        @Override
        public AsyncFuture<List<String>> serializeKeyToHex(
            final BackendKey key
        ) {
            return delegate.serializeKeyToHex(key);
        }

        @Override
        public AsyncFuture<List<BackendKey>> deserializeKeyFromHex(
            final String key
        ) {
            return delegate.deserializeKeyFromHex(key);
        }

        @Override
        public AsyncFuture<Void> deleteKey(
            final BackendKey key, final QueryOptions options
        ) {
            return delegate.deleteKey(key, options);
        }

        @Override
        public AsyncFuture<Long> countKey(
            final BackendKey key, final QueryOptions options
        ) {
            return delegate.countKey(key, options);
        }

        @Override
        public AsyncFuture<MetricCollection> fetchRow(
            final BackendKey key
        ) {
            return delegate.fetchRow(key);
        }

        @Override
        public AsyncObservable<MetricCollection> streamRow(
            final BackendKey key
        ) {
            return delegate.streamRow(key);
        }

        @Override
        public boolean isReady() {
            return delegate.isReady();
        }

        @Override
        public Groups groups() {
            return delegate.groups();
        }
    }
}

File: heroic-dist/src/test/java/com/spotify/heroic/LoggingMetricModule.kt
package com.spotify.heroic

import com.spotify.heroic.async.AsyncObservable
import com.spotify.heroic.common.Groups
import com.spotify.heroic.common.Statistics
import com.spotify.heroic.dagger.PrimaryComponent
import com.spotify.heroic.instrumentation.OperationsLog
import com.spotify.heroic.lifecycle.LifeCycle
import com.spotify.heroic.metric.*
import eu.toolchain.async.AsyncFuture
import io.opencensus.trace.Span
import java.util.function.Consumer

data class LoggingMetricModule(
    val backingStore: MetricModule, val opLog: OperationsLog
) : MetricModule {

    override fun module(
        primary: PrimaryComponent, backend: MetricModule.Depends, id: String
    ): MetricModule.Exposed {
        val backingStoreModule: MetricModule.Exposed = backingStore.module(primary, backend, id)

        return object : MetricModule.Exposed {
            override fun life(): LifeCycle {
                return backingStoreModule.life()
            }

            override fun backend(): MetricBackend {
                val delegate: MetricBackend = backingStoreModule.backend()

                return LoggingMetricBackend(delegate)
            }
        }
    }

    private inner class LoggingMetricBackend(val delegate: MetricBackend) : MetricBackend {

        override fun getStatistics(): Statistics {
            return delegate.statistics
        }

        override fun configure(): AsyncFuture<Void> {
            return delegate.configure()
        }

        override fun write(write: WriteMetric.Request): AsyncFuture<WriteMetric> {
            opLog.registerWriteRequest()
            return delegate.write(write).directTransform { result ->
                opLog.registerWriteComplete()
                result
            }
        }

        override fun fetch(
            request: FetchData.Request,
            watcher: FetchQuotaWatcher,
            metricsConsumer: Consumer<MetricReadResult>,
            span: Span
        ): AsyncFuture<FetchData.Result> {
            return delegate.fetch(request, watcher, metricsConsumer, span)
        }

        override fun listEntries(): Iterable<BackendEntry> {
            return delegate.listEntries()
        }

        override fun streamKeys(
            filter: BackendKeyFilter, options: QueryOptions
        ): AsyncObservable<BackendKeySet> {
            return delegate.streamKeys(filter, options)
        }

        override fun streamKeysPaged(
            filter: BackendKeyFilter, options: QueryOptions, pageSize: Long
        ): AsyncObservable<BackendKeySet> {
            return delegate.streamKeysPaged(filter, options, pageSize)
        }

        override fun serializeKeyToHex(key: BackendKey): AsyncFuture<List<String>> {
            return delegate.serializeKeyToHex(key)
        }

        override fun deserializeKeyFromHex(key: String): AsyncFuture<List<BackendKey>> {
            return delegate.deserializeKeyFromHex(key)
        }

        override fun deleteKey(key: BackendKey, options: QueryOptions): AsyncFuture<Void> {
            return delegate.deleteKey(key, options)
        }

        override fun countKey(key: BackendKey, options: QueryOptions): AsyncFuture<Long> {
            return delegate.countKey(key, options)
        }

        override fun fetchRow(key: BackendKey): AsyncFuture<MetricCollection> {
            return delegate.fetchRow(key)
        }

        override fun streamRow(key: BackendKey): AsyncObservable<MetricCollection> {
            return delegate.streamRow(key)
        }

        override fun isReady(): Boolean {
            return delegate.isReady
        }

        override fun groups(): Groups {
            return delegate.groups()
        }
    }
}
